awscommons/v2/ecs.go

```diff
@@ -0,0 +1,201 @@
+			Status:             "DRAINING",
+		}
+		_, err := client.UpdateContainerInstancesState(opts.Context, input)
+		if err != nil {
+			errList.AddError(err)
+			logger.Errorf("Encountered error starting to drain container instances in batch %d: %s", batchIdx, err)
+			logger.Errorf("Container Instance ARNs: %s", strings.Join(batchedArnList, ","))
+			continue
+		}
+
+		logger.Infof("Started draining %d container instances from batch %d", len(batchedArnList), batchIdx)
+	}
+
+	if !errList.IsEmpty() {
+		return errors.WithStackTrace(errList)
+	}
+	logger.Infof("Successfully started draining all %d container instances", len(containerInstanceArns))
+	return nil
+}
+
+// WaitForContainerInstancesToDrain waits until there are no more ECS Tasks running on any of the ECS container
+// instances. Batches container instances in groups of 100 because of AWS API limitations.
+func WaitForContainerInstancesToDrain(opts *Options, clusterName string, containerInstanceArns []string, start time.Time, timeout time.Duration, maxRetries int, sleepBetweenRetries time.Duration) error {
+	client, err := NewECSClient(opts)
+	if err != nil {
+		return err
+	}
+
+	logger := logging.GetProjectLogger()
+	logger.Infof("Checking if all ECS Tasks have been drained from the ECS Container Instances in Cluster %s.", clusterName)
+
+	batchSize := 100
+	numBatches := int(math.Ceil(float64(len(containerInstanceArns) / batchSize)))
+
+	err = retry.DoWithRetry(
+		logger.Logger,
+		"Wait for Container Instances to be Drained",
+		maxRetries, sleepBetweenRetries,
+		func() error {
+			responses := []*ecs.DescribeContainerInstancesOutput{}
+			for batchIdx, batchedArnList := range collections.BatchListIntoGroupsOf(containerInstanceArns, batchSize) {
+				batchedArns := aws.StringSlice(batchedArnList)
+
+				logger.Infof("Fetching description of batch %d/%d of ECS Instances in Cluster %s.", batchIdx, numBatches, clusterName)
+				input := &ecs.DescribeContainerInstancesInput{
+					Cluster:            aws.String(clusterName),
+					ContainerInstances: aws.ToStringSlice(batchedArns),
+				}
+				result, err := client.DescribeContainerInstances(opts.Context, input)
+				if err != nil {
+					return errors.WithStackTrace(err)
+				}
+				responses = append(responses, result)
+			}
+
+			// If we exceeded the timeout, halt with error.
+			if timeoutExceeded(start, timeout) {
+				return retry.FatalError{Underlying: fmt.Errorf("maximum drain timeout of %s seconds has elapsed and instances are still draining", timeout)}
+			}
+
+			// Yay, all done.
+			if drained, _ := allInstancesFullyDrained(responses); drained == true {
+				logger.Infof("All container instances have been drained in Cluster %s!", clusterName)
+				return nil
+			}
+
+			// If there's no error, retry.
+			if err == nil {
+				return errors.WithStackTrace(fmt.Errorf("container instances still draining"))
+			}
+
+			// Else, there's an error, halt and fail.
+			return retry.FatalError{Underlying: err}
```

Comment:
This `err` check is a tautology, since `err` is only set within the for loop above, and you immediately return `err` if there is an error. So it will always be `nil` at this point.

You can reduce this to return the `container instances still draining` error.

Reply:
Ah, I see. I think the best thing is to retry in all other cases. If timeout exceeded, exit with error. If drained, return nil. All other cases, retry.